spring-cloud-starter-stream-collector-metrics/src/main/java/org/springframework/cloud/stream/app/metrics/collector/MetricsAggregator.java

```diff
@@ -0,0 +1,122 @@
+ * This class handles conversion, update of deltas for the sendCount, as well as invalidation of stale entries
+ * @author Vinicius Carvalho
+ */
+@Component
+public class MetricsAggregator {
+
+	public static final String OUTPUT_METRIC_NAME = "integration.channel.output.sendRate.mean";
+	public static final String INPUT_METRIC_NAME = "integration.channel.input.sendRate.mean";
+
+	private Lock lock = new ReentrantLock();
+
+	private Cache<String,Stream> streamCache;
+
+	private Cache<String,ApplicationMetrics> rawCache;
+
+	public MetricsAggregator(@Qualifier("streamCache") Cache<String,Stream> streamCache, @Qualifier("rawCache") Cache<String,ApplicationMetrics> rawCache){
+		this.streamCache = streamCache;
+		this.rawCache = rawCache;
+	}
+
+
+	@StreamListener(Sink.INPUT)
+	public void receive(ApplicationMetrics metrics) {
+		this.rawCache.put(metrics.getName(),metrics);
+		Stream stream = convert(metrics);
+		Application application = stream.getApplications().get(0);
+		Instance instance = application.getInstances().get(0);
+		try{
+			lock.lock();
+			Stream cached = this.streamCache.getIfPresent(stream.getName());
+			if(cached == null){
+				this.streamCache.put(stream.getName(),stream);
+				return;
+			}
+			if(cached.getApplications().size() == 0){
+				cached.setApplications(stream.getApplications());
+				return;
+			}
+			int applicationIndex = cached.getApplications().indexOf(application);
+			if(applicationIndex >=0 ){
+				Application cachedApplication = cached.getApplications().get(applicationIndex);
+				if(cachedApplication.getInstances().size() == 0){
+					cachedApplication.getInstances().add(instance);
+					return;
+				}
+				int idx = cachedApplication.getInstances().indexOf(instance);
+				if(idx >= 0){
+					Instance cachedInstance = cachedApplication.getInstances().get(idx);
+					cachedInstance.updateOutgoingRate(instance.getOutgoingRate());
+					cachedInstance.updateIncomingRate(instance.getIncomingRate());
+				}else{
+					cachedApplication.getInstances().add(instance);
+				}
+			}else{
+				cached.getApplications().add(application);
+			}
+		}catch (Exception e){
+
+		}finally {
+			lock.unlock();
+		}
+	}
+
+
+	private Stream convert(ApplicationMetrics metrics){
+		Assert.notNull(metrics.getProperties().get(ApplicationMetrics.STREAM_NAME), "Missing STREAM_NAME from metrics properties");
+		Assert.notNull(metrics.getProperties().get(ApplicationMetrics.APPLICATION_NAME), "Missing APPLICATION_NAME from metrics properties");
+		Assert.notNull(metrics.getProperties().get(ApplicationMetrics.APPLICATION_GUID), "Missing APPLICATION_GUID from metrics properties");
+
+		Stream stream = new Stream((String)metrics.getProperties().get(ApplicationMetrics.STREAM_NAME));
+		Application application = new Application((String)metrics.getProperties().get(ApplicationMetrics.APPLICATION_NAME));
+
+		Double incomeRate = YANUtils.cast(metrics.getProperties().get(INPUT_METRIC_NAME),Double.class).orElse(0.0);
```

Comment:
Not sure I understand this but metrics are not coming in with properties? I don't see any processing of metrics itself of format from docs
```
{
  "name" : "application",
  "instanceIndex" : 0,
  "createdTime" : "2017-03-14T13:55:38.547Z",
  "properties" : {
    "java.specification.version" : "1.8"
  }
  "metrics" : [ {
    "name" : "mem",
    "value" : 170757.0,
    "timestamp" : "2017-03-14T09:55:38.547Z"
  } ]
}
```